middlewares/retry.go

```diff
@@ -56,8 +63,34 @@ func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
 	}
 }
 
-func isNetworkError(status int) bool {
-	return status == http.StatusBadGateway || status == http.StatusGatewayTimeout
+// netErrorCtxKey is a custom type that is used as key for the context.
+type netErrorCtxKey string
+
+// defaultNetErrCtxKey is the actual key which value is used to record network errors.
+var defaultNetErrCtxKey netErrorCtxKey = "NetErrCtxKey"
+
+// NetErrorRecorder is an interface to record net errors.
+type NetErrorRecorder interface {
+	// Record can be used to signal the retry middleware that an network error happened
+	// and therefore the request should be retried.
+	Record(ctx context.Context)
+}
+
+// NetErrorRecorderFunc is an util to be able to implement the NetErrorRecorder without the need of a struct.
+type NetErrorRecorderFunc func(ctx context.Context)
+
+// Record is there to satisfy the NetErrorRecorder interface on a NetErrorRecorderFunc.
+func (f NetErrorRecorderFunc) Record(ctx context.Context) {
+	f(ctx)
+}
+
+// DefaultRecordNetError is recording network errors by setting the context value for the defaultNetErrCtxKey to true.
+func DefaultRecordNetError(ctx context.Context) {
+	val := ctx.Value(defaultNetErrCtxKey)
+
+	if netErrorOccurred, ok := val.(*bool); ok {
```

Comment:
I renamed `ok` now to `isBoolPointer`. Not sure how to improve readability in another way.